crates/noirc_evaluator/src/ssa_refactor/acir_gen/mod.rs

```diff
@@ -363,6 +366,55 @@ impl Context {
             _ => todo!(),
         }
     }
+
+    /// Operands in a binary operation are checked to have the same type.
+    ///
+    /// In Noir, binary operands should have the same type due to the language
+    /// semantics.
+    ///
+    /// There are some edge cases to consider:
+    /// - Constants are not explicitly type casted, so we need to check for this and
+    /// return the type of the other operand, if we have a constant.
+    /// - 0 is not seen ad `Field 0` but instead as `Unit 0`
+    /// TODO: The latter seems like a bug, if we cannot differentiate between a function returning
+    /// TODO nothing and a 0.
+    ///
+    /// TODO: This constant coercion should ideally be done in the type checker.
+    fn type_of_binary_operation(&self, binary: &Binary, dfg: &DataFlowGraph) -> Type {
+        let lhs_type = dfg.type_of_value(binary.lhs);
+        let rhs_type = dfg.type_of_value(binary.rhs);
+
+        match (lhs_type, rhs_type) {
+            // Function type should not be possible, since all functions
+            // have been inlined.
+            (Type::Function, _) | (_, Type::Function) => {
+                unreachable!("all functions should be inlined")
+            }
+            // Unit type currently can mean a 0 constant, so we return the
+            // other type.
+            (typ, Type::Unit) | (Type::Unit, typ) => typ,
+            // If either side is a constant then, we coerce into the type
+            // of the other operand
+            (Type::Numeric(NumericType::NativeField), typ)
+            | (typ, Type::Numeric(NumericType::NativeField)) => typ,
+            // This should not be possible, if we are adding a reference
+            // it will be to a constant. This may not be true for dynamic arrays.
+            (Type::Numeric(_), Type::Reference) | (Type::Reference, Type::Numeric(_)) => {
+                unreachable!(
+                    "operations involving references should have a constant as the other operand."
+                )
+            }
+            (Type::Numeric(lhs_type), Type::Numeric(rhs_type)) => {
+                assert_eq!(
+                    lhs_type, rhs_type,
+                    "lhs and rhs types in a binary operation are always the same"
+                );
+                Type::Numeric(lhs_type)
+            }
+            (Type::Reference, Type::Reference) => Type::Reference,
```

Comment:
This is a good point -- I was going back and forth whether to include references since it is handled in mod.rs and not in `acir_variable`. I will make it unreachable